revisions/announcer/announcer.go

```diff
@@ -0,0 +1,293 @@
+	URL        string
+	RemoteName string
+	BranchName string
+	Depth      int
+	Tags       git.TagMode
+	EpochReferenceIterFactory
+	agit.Git
+}
+
+type gitRemoteAnnouncer struct {
+	repo agit.Repository
+	cfg  *GitRemoteAnnouncerConfig
+}
+
+// NewGitRemoteAnnouncer produces an Announcer that is bound to an agit.Repository.
+func NewGitRemoteAnnouncer(cfg GitRemoteAnnouncerConfig) (Announcer, error) {
+	a := &gitRemoteAnnouncer{
+		cfg: &cfg,
+	}
+
+	// Initialize freshness and repo according to cfg.
+	err := a.Reset()
+	if err == nil && a.repo == nil {
+		err = errNilRepo
+	}
+	if err != nil {
+		log.Printf("ERRO: Failed to construct git remote announcer: %v", err)
+		return nil, err
+	}
+
+	return a, err
+}
+
+// GetRevisions returns as complete a list of revisions as possible given current state. It will not fetch new revisions, but it will search for newer epochal revisions than a previous invocation (if any) based on current local repository state.
+func (a *gitRemoteAnnouncer) GetRevisions(epochs map[epoch.Epoch]int, limits Limits) (map[epoch.Epoch][]agit.Revision, error) {
+	// Create copy of epochs; local copy will be mutated.
+	es := make(map[epoch.Epoch]int)
+	for e, i := range epochs {
+		es[e] = i
+	}
+
+	// Initialize iterator according to config.
+	iter, err := a.cfg.EpochReferenceIterFactory.GetIter(a.repo, limits)
+	if err != nil {
+		log.Printf("ERRO: Failed to initialize reference iterator: %v", err)
+		return nil, err
+	}
+
+	revs := make(map[epoch.Epoch][]agit.Revision)
+	numChanges := 0
+	for e, i := range es {
+		revs[e] = make([]agit.Revision, 0, i)
+		numChanges += i
+	}
+
+	if numChanges == 0 {
+		return nil, errVacuousEpochs
+	}
+
+	// iter presents potential revisions in reverse chronological order.
+	// Scan for first epochal changes between nextTime and prevTime.
+	numChangesFound := 0
+	prevTime := limits.Now
+	for ref, err := iter.Next(); ref != nil && err == nil; ref, err = iter.Next() {
+		c, err := a.repo.CommitObject(ref.Hash())
+		if err != nil {
+			log.Printf("WARN: Failed to locate commit for PR tag: %s; skipping...", ref.Name())
+			continue
+		}
+		nextTime := c.Committer.When
+
+		// Check for epochal change against every epoch.
+		for e, i := range es {
+			if i == 0 {
+				continue
+			}
+			if e.IsEpochal(nextTime, prevTime) {
+				numChangesFound++
+				es[e]--
+
+				revs[e] = append(revs[e], agit.RevisionData{
+					Hash:       c.Hash,
+					CommitTime: nextTime,
+				})
+
+				if numChangesFound == numChanges {
+					break
+				}
+			}
+		}
+
+		if numChangesFound == numChanges {
+			break
+		}
+		prevTime = nextTime
+	}
+
+	// Surface error if not all epochs have a revision.
+	if numChangesFound != numChanges {
+		return revs, errNotAllEpochsConsumed
+	}
+
+	return revs, nil
+}
+
+// Update performs a fetch on the underlying repository. Subsequent calls to GetRevisions() will incorporate any newly fetched revisions.
+func (a *gitRemoteAnnouncer) Update() (err error) {
```

Comment:
Nit: Given our dependency on git, naming the function to match the git behaviour is probably easier on the reader.
i.e. call this Fetch

Reply:
Done.